types/tx.go

```diff
@@ -92,14 +94,213 @@ func (txs Txs) ToSliceOfBytes() [][]byte {
+		if i+1 < len(allCopy) && bytes.Equal(cur, allCopy[i+1]) {
+			return fmt.Errorf("found duplicate transaction with hash: %x", cur.Hash())
+		}
+	}
+
+	// create copies of each of the action-specific indexes so that order of the original
+	// indexes can be preserved.
+	addedCopy := sortedCopy(t.added)
+	removedCopy := sortedCopy(t.removed)
+	unmodifiedCopy := sortedCopy(t.unmodified)
+
+	// make a defensive copy of otxs so that the order of
+	// the caller's data is not altered.
+	otxsCopy := sortedCopy(otxs)
+
+	if ix, ok := containsAllTxs(otxsCopy, unmodifiedCopy); !ok {
+		return fmt.Errorf("new transaction incorrectly marked as removed, transaction hash: %x", unmodifiedCopy[ix].Hash())
+	}
+
+	if ix, ok := containsAllTxs(otxsCopy, removedCopy); !ok {
+		return fmt.Errorf("new transaction incorrectly marked as removed, transaction hash: %x", removedCopy[ix].Hash())
+	}
+	if ix, ok := containsAnyTxs(otxsCopy, addedCopy); ok {
+		return fmt.Errorf("existing transaction incorrectly marked as added, transaction hash: %x", addedCopy[ix].Hash())
+	}
+	return nil
+}
+
+func sortedCopy(txs Txs) Txs {
+	cp := make(Txs, len(txs))
+	copy(cp, txs)
+	sort.Sort(cp)
+	return cp
+}
+
+// containsAnyTxs checks that list a contains one of the transactions in list
+// b. If a match is found, the index in b of the matching transaction is returned.
+// Both lists must be sorted.
+func containsAnyTxs(a, b []Tx) (int, bool) {
+	aix, bix := 0, 0
+
+nextA:
+	for ; aix < len(a); aix++ {
+		for bix < len(b) {
+			switch bytes.Compare(b[bix], a[aix]) {
+			case 0:
+				return bix, true
+			case -1:
+				bix++
+				// we've reached the end of b, and the last value in b was
+				// smaller than the value under the iterator of a.
+				// a's values never get smaller, so we know there are no more matches
+				// in the list. We can terminate the iteration here.
+				if bix == len(b) {
+					return -1, false
+				}
+			case 1:
+				continue nextA
+			}
+		}
+	}
+	return -1, false
+}
+
+// containsAllTxs checks that super contains all of the transactions in the sub
+// list. If not all values in sub are present in super, the index in sub of the
+// first Tx absent from super is returned.
+func containsAllTxs(super, sub []Tx) (int, bool) {
+	// The following iteration assumes sorted lists.
+	// The checks ensure that all the values in the sorted sub list are present in the sorted super list.
+	//
+	// We compare the value under the sub iterator to the value
+	// under the super iterator. If they match, we advance the
+	// sub iterator one position. If they don't match, then the value under
+	// the sub iterator should be greater.
+	// If it is not, then there is a value in the the sub list that is not present in the
+	// super list.
+	subIx := 0
+	for _, cur := range super {
+		if subIx == len(sub) {
+			return -1, true
+		}
+		switch bytes.Compare(sub[subIx], cur) {
+		case 0:
+			subIx++
+		case -1:
+			return subIx, false
+		}
```

Comment:
What you have here is fine, but I feel like this might be simpler if we use the order directly rather than as a merge property in this context:

```go
func containsAll(super, sub Txs) (int, bool) {
	for i, cur := range sub {
		n := sort.Search(len(super), func(i int) bool {
			return bytes.Compare(cur, super[i]) <= 0
		})
		if n == len(super) || !bytes.Equal(cur, super[n]) {
			return i, false
		}
	}
	return -1, true
}
```

YMMV, but I think a little easier to be sure about (the index tracking always tingles my spider senses, though here I'm convinced it's correct)

Reply:
Fair enough. I was pleased with my little implementation, but it did take a while to convince myself it was correct. I'm happy to eat the mild runtime hit if it means this is easier for people to reason about in the future.